### PR TITLE
Devrel 1215/sync languages

### PR DESCRIPTION
**Pull request by IvanKiral**
### Motivation

Which issue does this fix? Fixes #`issue number`

If no issue exists, what is the fix or new feature? Were there any reasons to fix/implement things that are not obvious?

### Checklist

- [ ] Code follows coding conventions held in this repo
- [ ] Automated tests have been added
- [ ] Tests are passing
- [ ] Docs have been updated (if applicable)
- [ ] Temporary settings (e.g. variables used during development and testing) have been reverted to defaults

### How to test

If manual testing is required, what are the steps?

### CODE REVIEW

File: src/modules/sync/diff.ts [hunk 1]
```diff
@@ -117,3 +130,55 @@ const createDiffModel = <Entity extends Readonly<{ codename: string }>>(
...
+    languageHandler,
+  );
+
+  /**
+   * Manually add an operation to change the codename of the default language in the target environment
+   * to match the codename of the default language in the source environment.
+   */
+
+  languageDiffModel.updated.set(targetDefaultLanguageCodename, [
+    ...languageDiffModel.updated.get(targetDefaultLanguageCodename) ?? [],
```
[the remark this answers is not in the document]
Comment (JiriLojda) [reply src/modules/sync/diff.ts:159]: Also, can you change the data for integration tests so the change of the default language's codename is there? So we test this corner case.

File: src/modules/sync/diff/language.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+export const languageHandler: Handler<LanguageSyncModel> = makeObjectHandler({
+  name: baseHandler,
+  is_active: baseHandler,
+  is_default: constantHandler,
+  fallback_language: optionalHandler(makeLeafObjectHandler({
+    codename: (source, target) => source === target,
+  })),
+});
+
+export const wholeLanguageHandler: Handler<ReadonlyArray<LanguageSyncModel>> = makeArrayHandler(
```
Comment (JiriLojda) [inline src/modules/sync/diff/language.ts:21]: Once you rebase on the current `main`, use the new whole objects handler.

File: src/modules/sync/generateSyncModel.ts [hunk 3]
```diff
@@ -145,6 +165,14 @@ export const saveSyncModel = async (params: SaveModelParams) => {
     path.resolve(folderName, assetFoldersFileName),
     JSON.stringify(finalModel.assetFolders, null, 2),
   );
+  await fsPromises.writeFile(
+    path.resolve(folderName, collectionsFileName),
+    JSON.stringify(finalModel.collections, null, 2),
+  );
+  await fsPromises.writeFile(
+    path.resolve(folderName, languagesFileName),
+    JSON.stringify(finalModel.collections, null, 2),
```
Comment (JiriLojda) [inline src/modules/sync/generateSyncModel.ts:174]: C&P all the way 😄

File: src/modules/sync/diff.ts [hunk 4]
```diff
@@ -117,3 +130,55 @@ const createDiffModel = <Entity extends Readonly<{ codename: string }>>(
       },
       { added: [], updated: new Map(), deleted: new Set() },
     );
+
+const getLanguageDiffModel = (params: DiffParams) => {
```
Comment (JiriLojda) [inline src/modules/sync/diff.ts:134]: What about adding a unit test on the whole `diff` function that would test that the change of the default language's codename is handled properly here?

Comment (JiriLojda) [reply src/modules/sync/diff.ts:134]: Also, you don't need to accept the whole `DiffParams` here. Just the source and target languages.

**Review by JiriLojda — CHANGES_REQUESTED** [review]
Update the test data export zips, please. Just to make sure that they are up-to-date with all the latest changes. We updated them in all our recent branches and I am not certain that you have the latest snapshot here.

File: src/modules/sync/sync/languages.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+const deleteLanguage = (client: ManagementClient, codename: string) => {
+  const randomUuid = createUuid();
+  return client
+    .modifyLanguage()
+    .byLanguageCodename(codename).withData([
+      /**
+       * languages cannot be deleted, instead they are deactivated
+       * and their name and codename are both populated with first 8 chars of their ID
+       * (name and codename have a limit of 25 characters).
```
[the remark this answers is not in the document]
Comment (IvanKiral) [reply src/modules/sync/sync/languages.ts:68]: We will need to update the whole readme after we add all of the entity.. It would be better to adjust it slightly before the release as it might confuse users with the newly added features that are not released yeat

File: src/modules/sync/sync/languages.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+});
+
+const createReplaceIsActiveOperation = (isActive: boolean): LanguageModels.IModifyLanguageData => ({
+  op: "replace",
+  property_name: "is_active",
+  value: isActive,
+});
+
+const operationsToOrdNumb = (operations: LanguageModels.IModifyLanguageData[]) =>
+  operations.find(op => op.property_name === "codename") !== undefined ? 0 : 100;
```
[the remark this answers is not in the document]
Comment (IvanKiral) [reply src/modules/sync/sync/languages.ts:149]: diff handler can't... but I can 
image
:D

this is due to change of codename of default language

File: src/modules/sync/diff.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (IvanKiral) [reply src/modules/sync/diff.ts:166]: I absolutely can't as it will mess up syncContent tests :/

File: src/modules/sync/sync/languages.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    .returnType<LanguageModels.IModifyLanguageData>()
+    .with({ op: "replace" }, (op) => {
+      const splittedPath = op.path.split("/");
+      const propertyName = splittedPath[splittedPath.length - 1];
+      if (!propertyName) {
+        throw new Error(`Could not find property name in path ${op.path}`);
+      }
+
+      if (!isValidLanguagePropertyName(propertyName)) {
+        throw new Error(`Property name ${propertyName} should is not a correct value`);
```
[the remark this answers is not in the document]
Comment (JiriLojda) [reply src/modules/sync/sync/languages.ts]: The check is redundant as well. 😉
